source/material_model/averaging.cc

```diff
@@ -0,0 +1,449 @@
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = average;
+        }
+      else if (averaging_operation ==  "pick largest")
+        {
+          double max = -std::numeric_limits<double>::max();
+          for (unsigned int i=0; i<N; ++i)
+            max = std::max(max, values_out[i]);
+
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = max;
+          //break;
+        }
+      else if (averaging_operation == "nwd arithmetic average")
+        {
+          //initialize variables
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+
+          // determine the maximum distance between all the points
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  if (position[i].distance(position[j]) > max_distance)
+                    max_distance = position[i].distance(position[j]);
+                }
+            }
+
+          // apply the averaging to the values
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
+
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  double distance = position[i].distance(position[j])/max_distance;
+                  double weight = alfad * ((1 + 3 * (distance / bell_shape_limit)) * (1 - (distance / bell_shape_limit)) * (1 - (distance / bell_shape_limit)) * (1 - (distance / bell_shape_limit)));
+                  if (distance / bell_shape_limit > 1)
+                    weight = 0;
+                  sum_value += weight * values_out[j];
+                  sum_weights += weight;
+                }
+
+              const double average = sum_value / sum_weights;
+
+              temp_values[i] = average;
+            };
+          for (unsigned int i = 0; i<N; ++i)
+            {
+              values_out[i] = temp_values[i];
+            }
+        }
+      else if (averaging_operation == "nwd harmonic average")
+        {
+          // if one of the values is zero, the average is 0.0
+          for (unsigned int i=0; i<N; ++i)
+            if (values_out[i] == 0.0)
+              {
+                for (unsigned int j=0; j<N; ++j)
+                  values_out[j] = 0.0;
+                return;
+              }
+
+          //initialize variables
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+
+          // determine the maximum distance between all the points
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  if (position[i].distance(position[j]) > max_distance)
+                    max_distance = position[i].distance(position[j]);
```

Comment:
You can write this more compactly as
```
  max_distance = std::max (max_distance,
                                            position[i].distance(position[j]));
```
This way you also don't have to compute the distance twice.